rotkehlchen/chain/evm/decoding/odos/common.py

```diff
@@ -0,0 +1,202 @@
+                    address=tx_log.address,
+                    chain_id=self.evm_inquirer.chain_id,
+                    token_type=EvmTokenKind.ERC20,
+                )) not in output_tokens)
+            ):
+                continue
+
+            # can directly initialize EvmToken here because it's in output_tokens which were get_or_created above  # noqa: E501
+            asset = EvmToken(identifier)
+            if hex_or_bytes_to_address(tx_log.topics[2]) == self.router_address:
+                router_holding_amount[asset] += asset_normalized_value(amount=hex_or_bytes_to_int(tx_log.data), asset=asset)  # noqa: E501
+
+            elif hex_or_bytes_to_address(tx_log.topics[1]) == self.router_address:
+                router_holding_amount[asset] -= asset_normalized_value(amount=hex_or_bytes_to_int(tx_log.data), asset=asset)  # noqa: E501
+
+        if self.native_currency.identifier in output_tokens:
+            try:  # download (if not there) and find all native token transfers from/to the router
+                internal_native_ins = self.evm_txns.get_and_ensure_internal_txns_of_parent_in_db(
+                    tx_hash=context.transaction.tx_hash,
+                    to_address=self.router_address,
+                    user_address=sender,
+                )
+                internal_native_outs = self.evm_txns.get_and_ensure_internal_txns_of_parent_in_db(
+                    tx_hash=context.transaction.tx_hash,
+                    from_address=self.router_address,
+                    to_address=sender,
+                    user_address=sender,
+                )
+            except RemoteError as e:
+                log.error(f'Failed to get internal transactions for {self.label} swap {context.transaction} due to {e!s}')  # noqa: E501
+            else:
+                for internal_native_in in internal_native_ins:
+                    router_holding_amount[self.native_currency] += asset_normalized_value(amount=internal_native_in.value, asset=self.native_currency)  # noqa: E501
+                for internal_native_out in internal_native_outs:
+                    router_holding_amount[self.native_currency] -= asset_normalized_value(amount=internal_native_out.value, asset=self.native_currency)  # noqa: E501
+
+        return [
+            self.base.make_event_next_index(
+                tx_hash=context.transaction.tx_hash,
+                timestamp=context.transaction.timestamp,
+                event_type=HistoryEventType.SPEND,
+                event_subtype=HistoryEventSubType.FEE,
+                asset=held_asset,
+                balance=Balance(amount=amount),
+                location_label=sender,
+                notes=f'Spend {amount} {held_asset.symbol_or_name()} as an {self.label} fee',
+                counterparty=self.counterparties()[0].identifier,
+                address=self.router_address,
+            )
+            for held_asset, amount in router_holding_amount.items()
+            if amount != ZERO
```

Comment:
what happens if the router_holding_amount is negative? I don't see any checks against this.